src/commands/plugin-search.php

```diff
@@ -2,116 +2,206 @@
+			->setHelp( "This command will output a list of Jetpack sites connected to the a8cteam51 account where a particular plugin is installed.\nThe search can be made for an exact match plugin slug, or\na general text search. Letter case is ignored in both search types.\nExample usage:\nplugin-search woocommerce\nplugin-search woo --partial\n" );
 
-	protected function execute( InputInterface $input, OutputInterface $output ) {
-		$plugin_slug = strtolower( $input->getArgument( 'plugin-slug' ) );
+		$this->addArgument( 'plugin-slug', InputArgument::REQUIRED, "The slug of the plugin to search for. This is an exact match against the plugin installation folder name,\nthe main plugin file name without the .php extension, and the Text Domain.\n" )
+			->addOption( 'partial', null, InputOption::VALUE_NONE, "Optional.\nUse for general text/partial match search. Using this option will also search the plugin Name field." );
+	}
 
-		$api_helper = new API_Helper;
+	/**
+	 * {@inheritDoc}
+	 */
+	protected function initialize( InputInterface $input, OutputInterface $output ): void {
+		maybe_define_console_verbosity( $output->getVerbosity() );
 
-		$output->writeln( '<info>Fetching list of sites...<info>' );
+		$this->plugin_slug = \strtolower( $input->getArgument( 'plugin-slug' ) );
+		$this->partial     = (bool) $input->getOption( 'partial' );
+	}
 
-		$sites = $api_helper->call_wpcom_api( 'rest/v1.1/jetpack-blogs/', array() );
+	/**
+	 * {@inheritDoc}
+	 */
+	protected function execute( InputInterface $input, OutputInterface $output ): int {
+		$partial_match_text = $this->partial ? 'partial' : 'exact';
+		$output->writeln( "<fg=magenta;options=bold>Searching through the WPCOM Jetpack sites for the plugin with the slug `$this->plugin_slug` ($partial_match_text match).</>" );
 
+		// Get the list of sites.
+		$sites = get_wpcom_jetpack_sites();
 		if ( empty( $sites ) ) {
 			$output->writeln( '<error>Failed to fetch sites.<error>' );
-			exit;
-		}
-
-		$site_list = array();
-		foreach ( $sites->blogs->blogs as $site ) {
-			$site_list[] = array( $site->userblog_id, $site->domain );
+			return 1;
 		}
-		$site_count = count( $site_list );
-
-		$output->writeln( "<info>{$site_count} sites found.<info>" );
-		$output->writeln( "<info>Checking each site for the plugin slug: {$plugin_slug}<info>" );
-
-		$progress_bar = new ProgressBar( $output, $site_count );
-		$progress_bar->start();
 
+		$output->writeln( '<fg=green;options=bold>Successfully fetched ' . count( $sites ) . ' Jetpack sites.</>', OutputInterface::VERBOSITY_VERBOSE );
+
+		// Get the list of plugins for each site.
+		$sites_plugins = WPCOM_API_Helper::call_api_concurrent(
+			\array_map(
+				static fn( $site ) => "/jetpack-blogs/$site->userblog_id/rest-api/?path=/jetpack/v4/plugins",
+				$sites
+			)
+		);
+		$sites_plugins = \array_combine(
+			\array_column( $sites, 'userblog_id' ),
+			$sites_plugins
+		);
```

Comment:
Yes. Instead of going through the sites one by one, this calls the WPCOM API for all of them at once ... basically sending out ±700 requests in one go. We had some issues with rate limiting last year, but we've since upped our calls timeout and it's been working well ever since.

That trick was already being used by the `remove-user` command (hence why we've encountered the rate-limiting problems last year already).